Implement KOOK provider

gehongyan opened the pull request:
This pull request contains the provider implementations for KOOK OAuth2.

The corresponding Issue is #753

gehongyan committed:
Add KOOK provider (#753)

martincostello commented on docs/kook.md:
```diff
@@ -0,0 +1,25 @@
+            options.ClientId = "my-client-id";
+            options.ClientSecret = "my-client-secret";
+
+            // The scope get_user_info is added by default
+            // Please make sure the scope get_user_info is enabled in your KOOK developer center
+            // Hence, whether to add the scope get_user_info here manually is optional
+            // options.Scope.Add("get_user_info");
```
If it's added by default in the provider, then adding it manually would be a no-op.

I would suggest that the commented out line of code would instead _remove_ it if a user didn't want to enable it in the KOOK developer centre.

gehongyan replied:
Thanks for reviewing.
Did you mean that the provider should not add a scope that can be disabled by developers in the KOOK developer center by default?
Or I could preserve adding get_user_info by default, but I should replace the reviewed code line with `options.Scope.Remove("get_user_info");` to tell developers how to disable the scope get_user_info?

martincostello replied:
Add a default scope is fine if that's what the average user would do with the provider, but the comment should demonstrate removing it in the case someone doesn't wish to use it.

It's just that the comment doesn't make much sense adding something that's already there.

gehongyan replied:
Got it!

martincostello commented on src/AspNet.Security.OAuth.Kook/AspNet.Security.OAuth.Kook.csproj:
```diff
@@ -0,0 +1,23 @@
+    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
+  </PropertyGroup>
+
+  <!-- TODO Remove once published to NuGet.org -->
+  <PropertyGroup>
+    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
+  </PropertyGroup>
```
The baseline needs to be 7.0.1, then after publishing we'd only remove `DisablePackageBaselineValidation` and not the baseline property.

```suggestion
  <PropertyGroup>
    <!-- TODO Remove once published to NuGet.org -->
    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
    <PackageValidationBaselineVersion>7.0.1</PackageValidationBaselineVersion>
  </PropertyGroup>
```

martincostello commented on src/AspNet.Security.OAuth.Kook/AspNet.Security.OAuth.Kook.csproj:
```diff
@@ -0,0 +1,23 @@
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <Description>ASP.NET Core security middleware enabling Kook authentication.</Description>
```
"Kook" or "KOOK"?

martincostello commented on src/AspNet.Security.OAuth.Kook/KookAuthenticationOptions.cs:
```diff
@@ -0,0 +1,37 @@
+        ClaimActions.MapJsonKey(KookAuthenticationConstants.Claims.OperatingSystem, "os");
+        ClaimActions.MapJsonKey(KookAuthenticationConstants.Claims.AvatarUrl, "avatar");
+        ClaimActions.MapJsonKey(KookAuthenticationConstants.Claims.BannerUrl, "banner");
+        ClaimActions.MapJsonKey(KookAuthenticationConstants.Claims.IsBot, "bot");
```
Is a user logging in through the OAuth flow ever going to actually be a bot?

`IsOnline` also seems like it wouldn't be particularly useful if it's only capturing the state at the exact moment the user logs in.

martincostello commented on src/AspNet.Security.OAuth.Kook/KookAuthenticationHandler.cs:
```diff
@@ -0,0 +1,89 @@
+        var principal = new ClaimsPrincipal(identity);
+        var user = payload.RootElement.GetProperty("data");
+
+        identity.AddClaim(new Claim(KookAuthenticationConstants.Claims.IsBanned, (user.GetProperty("status").GetInt32() == 10).ToString()));
```
I'm not a fan of the `10` magic number here. Could you promote it to an inline constant with a meaningful name?

martincostello replied:
In fact, I think it would be better to not be interpreting the API shape and creating virtual claims (like "Is banned?") here at all.

Users can implement logic like that for themselves using the OAuth events and the custom claims mapping actions.

Please remove this claim and the `IsBanned` constant.

martincostello reviewed (commented):
Some initial feedback, but I can't access the documentation from the computer/network I'm currently on so I'll need to review it separately to cross-reference with the implementation.

gehongyan committed:
Apply review suggestions on KOOK provider

Removed IsBot, IsOnline, IsBanned claims
Updated unit tests reflecting the changes of claims
Fixed incorrect naming in project description
Optimized documentation around default scopes

Reviewed-by: martincostello

gehongyan commented:
The commit fcafb83275c05bfb545084dd2ceef136d553a8e7 applied review suggestions from @martincostello 

- Removed `IsBot`, `IsOnline`, `IsBanned` claims
- Updated unit tests reflecting the changes of claims
- Fixed incorrect naming in project description
- Added `PackageValidationBaselineVersion` to project property
- Optimized documentation around default scopes

martincostello commented on docs/kook.md:
```diff
@@ -0,0 +1,25 @@
+            options.ClientId = "my-client-id";
+            options.ClientSecret = "my-client-secret";
+
+            // The scope get_user_info is added by default
+            // Please make sure the scope get_user_info is enabled in your KOOK developer center
+            // If you do not want to use the default scope indeed, try remove it by
```
```suggestion
            // The scope get_user_info is added by default.
            // Please make sure the scope get_user_info is enabled in your KOOK developer center.
            // If you do not want to use the default scope indeed, remove it as shown below:
```

martincostello reviewed (commented):
Other than one minor comment this looks good. I just still need to review the provider documentation against the implementation.

gehongyan committed:
Fix improper syntax in comments

Reviewed-by: martincostello

Co-authored-by: Martin Costello <martin@martincostello.com>

gehongyan commented:
Hi, @martincostello,
I noticed that you mentioned you are unable to access the hosted documentation site. I have updated the reference links of the source code managed on GitHub in issue #753. I will list them here as well.

The source code of the documentation can be found here:
> <https://github.com/kaiheila/api-docs> (_Simplified Chinese_ only)

Pages related to OAuth2 are listed here:
> OAuth2 Intro: <https://github.com/kaiheila/api-docs/blob/main/docs/zh-cn/oauth2.md> (_Simplified Chinese_ only)
> OAuth API: <https://github.com/kaiheila/api-docs/blob/main/docs/zh-cn/http/oauth.md> (_Simplified Chinese_ only)

KOOK is renamed from KaiHeiLa. The official repository is still managed in `kaiheila` namespace.

martincostello commented on src/AspNet.Security.OAuth.Kook/KookAuthenticationOptions.cs:
```diff
@@ -0,0 +1,35 @@
+        ClaimActions.MapJsonKey(ClaimTypes.Name, "username");
+        ClaimActions.MapJsonKey(ClaimTypes.MobilePhone, "mobile");
+        ClaimActions.MapJsonKey(KookAuthenticationConstants.Claims.IdentifyNumber, "identify_num");
+        ClaimActions.MapJsonKey(KookAuthenticationConstants.Claims.OperatingSystem, "os");
+        ClaimActions.MapJsonKey(KookAuthenticationConstants.Claims.AvatarUrl, "avatar");
+        ClaimActions.MapJsonKey(KookAuthenticationConstants.Claims.BannerUrl, "banner");
```
`os` and `banner` don't appear to be present in the documentation.

![image](https://user-images.githubusercontent.com/1439341/232286738-cb0dd406-0a6d-44b2-85f3-566fac067697.png)

gehongyan replied:
The documentation is maintained by officials and developers together. The content may be not up to date. I will commit some pull requests to fix the unrenewed content soon.

By testing the actual behaviors, the API returns the content as below:

![image](https://user-images.githubusercontent.com/21241496/232319762-4377904c-6c4b-4774-a8df-a5083c3673bf.png)

By testing via the Mvc.Client project, I could get the authorization as below:

![image](https://user-images.githubusercontent.com/21241496/232319855-e9ee31a1-cefa-43ee-9ac2-bdfda907f88f.png)

gehongyan replied:
Documentation fixture pull request has been made: kaiheila/api-docs#184

martincostello reviewed (commented):
Sorry for taking so long to review this. Just one question.

martincostello commented on src/AspNet.Security.OAuth.Kook/AspNet.Security.OAuth.Kook.csproj:
```diff
@@ -0,0 +1,24 @@
+  <!-- TODO Remove once published to NuGet.org -->
+  <PropertyGroup>
+    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
+    <PackageValidationBaselineVersion>7.0.1</PackageValidationBaselineVersion>
```
```suggestion
    <PackageValidationBaselineVersion>7.0.2</PackageValidationBaselineVersion>
```

martincostello committed:
Update package baseline

Bump baseline for next release.

martincostello committed:
Merge branch 'dev' into dev-kook
